worker/raft/queue/blockingqueue.go

```diff
@@ -0,0 +1,67 @@
+// Copyright 2021 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package queue
+
+import (
+	"context"
+	"time"
+
+	"github.com/juju/errors"
+)
+
+// Operation holds the operations that a queue can hold.
+type Operation struct {
+	Command []byte
+	Timeout time.Duration
+}
+
+// BlockingOpQueue holds the operations in a blocking queue. The purpose of this
+// queue is to allow the backing off of operations at the source of enqueueing,
+// and not at the consuming of the queue.
+type BlockingOpQueue struct {
+	queue chan Operation
+	err   chan error
+}
+
+// NewBlockingOpQueue creates a new BlockingOpQueue.
+func NewBlockingOpQueue() *BlockingOpQueue {
+	return &BlockingOpQueue{
+		queue: make(chan Operation),
+		err:   make(chan error),
+	}
+}
+
+// Enqueue will add an operation to the queue. As this is a blocking queue, any
+// additional enqueue operations will block and wait for subsequent operations
+// to be completed.
+// The design of this is to ensure that people calling this will have to
+// correctly handle backing off from enqueueing.
+func (q *BlockingOpQueue) Enqueue(ctx context.Context, op Operation) error {
```

Comment:
Creating a new context per enqueue operation just so we can encode a timeout seems [expensive](https://cs.opensource.google/go/go/+/refs/tags/go1.17.1:src/context/context.go;l=342-349) to me given the volume of calls that a large controller may perform.

Why not pass a context to the queue constructor (so we can shut it down when the worker dies) and replace the first argument with a timeout value that we can use as an extra case in the select block below?

Reply:
It's not per queue, it's per operation item. As each operation has to timeout, we need to ensure that we don't allow unbounded requests.

Reply:
> Creating a new context per enqueue operation just so we can encode a timeout seems expensive to me given the volume of calls that a large controller may perform.

This is premature optimisation as well.

Also storing contexts in structs is **strongly** discouraged: https://go.dev/blog/context-and-structs